### PR TITLE
[12.0][MIG] mrp_mto_with_stock

### PR DESCRIPTION
**Pull request by sergiocorato**
Standard migration

### CODE REVIEW

**Comment by rousseldenis** [comment]
@sergiocorato I think this PR includes irrelevant modules

**Comment by sergiocorato** [comment]
> @sergiocorato I think this PR includes irrelevant modules

~~ok, but I found a lot of issues open about mts-mto not working with subassembly ([this](https://github.com/OCA/manufacture/issues/253) and [this](https://github.com/OCA/manufacture/issues/317) and [last](https://github.com/OCA/stock-logistics-warehouse/issues/572)), and I found this bug too
Maybe there's some configuration issue?~~

**Comment by sergiocorato** [comment]
> > @sergiocorato I think this PR includes irrelevant modules
> 
> ok, but I found a lot of issues open about mts-mto not working with subassembly ([this](https://github.com/OCA/manufacture/issues/253) and [this](https://github.com/OCA/manufacture/issues/317) and [last](https://github.com/OCA/stock-logistics-warehouse/issues/572)), and I found this bug too
> Maybe there's some configuration issue?

@rousseldenis I saw the problem in commit now, I fix it

File: mrp_mto_with_stock/models/mrp_production.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            move_ids = copy.copy(production.move_raw_ids.ids)
+            for move in move_obj.browse(move_ids):
+                new_move = False
+                qty_to_procure = 0.0
+                if move.state in ('partially_available', 'confirmed') \
+                        and move.procure_method == 'make_to_stock' \
+                        and mto_with_no_move_dest_id and \
+                        self._mto_with_stock_condition(move):
+                    qty_to_procure = production.get_mto_qty_to_procure(move)
+                    if qty_to_procure > 0.0:
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py]: Use float_compare here

File: mrp_mto_with_stock/models/mrp_production.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        @returns True"""
+        # reserve all that is available (standard behaviour):
+        res = super(MrpProduction, self).action_assign()
+        # try to create procurements:
+        move_obj = self.env['stock.move']
+        for production in self:
+            warehouse = production.location_src_id.get_warehouse()
+            mto_with_no_move_dest_id = warehouse.mrp_mto_mts_forecast_qty
+            move_ids = copy.copy(production.move_raw_ids.ids)
+            for move in move_obj.browse(move_ids):
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py:35]: Why not using recordsets directly ?

File: mrp_mto_with_stock/models/mrp_production.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            qty_to_procure = 0.0
+            if not mto_with_no_move_dest_id:
+                # We have to split the move because we can't have
+                # a part of the move that have ancestors and not the
+                # other else it won't ever be reserved.
+                qty_to_procure = min(
+                    move.product_uom_qty -
+                    move.product_id.qty_available_not_res,
+                    move.product_uom_qty)
+                if 0.0 < qty_to_procure < move.product_uom_qty:
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py]: Use float_compare here

File: mrp_mto_with_stock/models/mrp_production.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                        'unit_factor': move.unit_factor * ratio,
+                    })
+                    move.write({
+                        'product_uom_qty':
+                            move.product_uom_qty - qty_to_procure,
+                        'unit_factor': move.unit_factor * (1 - ratio),
+                    })
+                    move._action_confirm()
+                    move._action_assign()
+                elif qty_to_procure > 0.0:
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py]: Use float_compare here

File: mrp_mto_with_stock/models/mrp_production.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    def get_mto_qty_to_procure(self, move):
+        self.ensure_one()
+        stock_location_id = move.location_id.id
+        move_location = move.with_context(location=stock_location_id)
+        virtual_available = move_location.product_id.virtual_available
+        qty_available = move.product_id.uom_id._compute_quantity(
+            virtual_available, move.product_uom)
+        draft_incoming_qty = self._get_incoming_qty_waiting_validation(move)
+        qty_available += draft_incoming_qty
+        if qty_available >= 0:
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py]: Please use float_compare here

File: mrp_mto_with_stock/models/mrp_production.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            return 0.0
+        else:
+            if abs(qty_available) < move.product_uom_qty:
```
Comment (rousseldenis) [inline mrp_mto_with_stock/models/mrp_production.py]: And float_compare too :smiley:

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some comments

File: mrp_mto_with_stock/models/mrp_production.py (hunk 1 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: perhaps `float_is_zero` ?

Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: done

File: mrp_mto_with_stock/models/mrp_production.py (hunk 4 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: done

File: mrp_mto_with_stock/models/mrp_production.py (hunk 5 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: done

File: mrp_mto_with_stock/models/mrp_production.py (hunk 6 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: done thanks :D

File: mrp_mto_with_stock/models/mrp_production.py (hunk 2 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py:35]: it comes from [this commit](https://github.com/OCA/manufacture/commit/f206743efef91a4c61b7540c3c2b828113c5ada6), I'm unable to find some conversation about

File: mrp_mto_with_stock/models/mrp_production.py (hunk 3 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: done

**Comment by kos94ok-3D** [comment]
Hi @sergiocorato.
Maybe easier:
1. add a constraint for a warehouse that disallow
use MTS+MTO with manufacture_steps == 'mrp_one_steps'
2. add new mto+mts rule with a lower sequence in "My Company: Pick components and then manufacture" route
As a result, it is easier configuration and not need to install mrp_mto_with_stock_purchase

**Comment by LoisRForgeFlow** [comment]
Hi @sergiocorato and thanks for contributing!

Please follow migration wiki to preserve commit history on migration: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#technical-method-to-migrate-a-module-from-110-to-120-branch.

**Comment by kos94ok-3D** [comment]
Hi @rousseldenis. What do you think about my comment above?

**Comment by eng-eslam-a-y** [comment]
Hello Guys, i was wondering when this module would be migrated to 12.0 and what is the problem with it maybe i could help.

**Comment by rousseldenis** [comment]
@sergiocorato Could you check travis ?

**Comment by sergiocorato** [comment]
> Hi @sergiocorato and thanks for contributing!
> 
> Please follow migration wiki to preserve commit history on migration: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#technical-method-to-migrate-a-module-from-110-to-120-branch.

hi @LoisRForgeFlow thanks, done

**Comment by ruuter** [comment]
Hi, any status update regarding this PR?

**Comment by pedroortega-punt** [comment]
@sergiocorato  What is the status of the migration of the mrp_mto_with_stock and mrp_mto_with_stock_pushase modules?

**Comment by sergiocorato** [comment]
@vipvalen I'm not working on it in this period

**Comment by pedroortega-punt** [comment]
@sergiocorato But it will continue some day or there is no forecast to contemplate the functionality of the mts_mto in manufacturing

**Comment by sergiocorato** [comment]
I'll check for a solution within some months

**Comment by andreampiovesana** [comment]
hi, is it ready to merge? what else missing?

**Comment by LoisRForgeFlow** [comment]
> hi, is it ready to merge? what else missing?

Reviews

**Review by desdelinux — COMMENTED** [review]
In the functional review I got an error.

Create a bill of material with a component marked Buy and Manufacturing MTO / MTS Locations. Add stock to the component, make sure the component has a supplier.

Create a sales order for the finished product for more quantity than can be manufactured with the component that is in stock (which is marked as Manufacture and Make to Order).

The necessary production orders and purchase orders will be properly created, but the quantity calculated to order in the purchase order is always double that needed.

That is, if you need to purchase 2 units of the component to close manufacturing, the purchase order will be made for 4 units.

Example:

BOM: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=13&action=335&model=mrp.bom&view_type=form&menu_id=176

Sale Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=23&action=387&model=sale.order&view_type=form&menu_id=227

Manufacturing Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=6&action=328&model=mrp.production&view_type=form&menu_id=176

 In the Mo it is required to buy 2 units of the mp component

Purcharse Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=10&action=366&model=purchase.order&view_type=form&menu_id=206

In the PO, a quote was generated for 4 units. 2 x Quantity required (This is a bug)

@sergiocorato @rousseldenis @kos94ok-3D @LoisRForgeFlow your review, please

**Comment by sergiocorato** [comment]
> In the functional review I got an error.
> 
> Create a bill of material with a component marked Buy and Manufacturing MTO / MTS Locations. Add stock to the component, make sure the component has a supplier.
> 
> Create a sales order for the finished product for more quantity than can be manufactured with the component that is in stock (which is marked as Manufacture and Make to Order).
> 
> The necessary production orders and purchase orders will be properly created, but the quantity calculated to order in the purchase order is always double that needed.
> 
> That is, if you need to purchase 2 units of the component to close manufacturing, the purchase order will be made for 4 units.
> 
> Example:
> 
> BOM: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=13&action=335&model=mrp.bom&view_type=form&menu_id=176
> 
> Sale Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=23&action=387&model=sale.order&view_type=form&menu_id=227
> 
> Manufacturing Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=6&action=328&model=mrp.production&view_type=form&menu_id=176
> 
> In the Mo it is required to buy 2 units of the mp component
> 
> Purcharse Order: http://3498726-411-8f5695.runbot1-3.odoo-community.org/web?debug=1#id=10&action=366&model=purchase.order&view_type=form&menu_id=206
> 
> In the PO, a quote was generated for 4 units. 2 x Quantity required (This is a bug)
> 
> @sergiocorato @rousseldenis @kos94ok-3D @LoisRForgeFlow your review, please

Thanks @desdelinux , pushed commit to fix duplicated procurement, removed parts no more used and closed #412 , awaiting test results (and possibly re-reviews :) )

**Comment by sergiocorato** [comment]
~Runbot claims an error on mrp_multi_level:~

~File "/home/odoo/build/OCA/manufacture/mrp_multi_level/models/product_mrp_area.py", line 179, in _compute_supply_method
rule = group_obj._get_rule(rec.product_id, proc_loc, values)
File "/home/odoo/odoo-12.0/addons/stock/models/stock_rule.py", line 353, in _get_rule
domain_company = ['|', ('company_id', '=', False), ('company_id', 'child_of', values['company_id'].ids)]
AttributeError: 'int' object has no attribute 'ids'~

**Comment by sergiocorato** [comment]
@rousseldenis @kos94ok-3D @andreampiovesana @LoisRForgeFlow @vipvalen @desdelinux @OCA/manufacturing-maintainers : for me it's complete, please review if interested

**Review by raistdev — APPROVED** [review]

**Review by andreampiovesana — APPROVED** [review]
Go

**Comment by LoisRForgeFlow** [comment]
@rousseldenis could you update your review?

**Review by rousseldenis — APPROVED** [review]

File: mrp_mto_with_stock/__manifest__.py [hunk 7]
```diff
@@ -7,9 +8,10 @@
     "summary": "Fix Manufacturing orders to pull from stock until qty is "
                "zero, and then create a procurement for them.",
     "author": "John Walsh, Eficent, Odoo Community Association (OCA)",
+    "maintainer": "Sergio Corato",
```
Comment (LoisRForgeFlow) [inline mrp_mto_with_stock/__manifest__.py]: @sergiocorato You should put your github login (`sergiocorato`) here, not your name.

File: mrp_mto_with_stock/models/mrp_production.py [hunk 8]
```diff
@@ -80,19 +88,8 @@ def _adjust_procure_method(self):
...
-                    new_move = move
-                else:
-                    # If we don't need to procure, we reserve the qty
-                    # for this move so it won't be available for others,
-                    # which would generate planning issues.
-                    move._action_confirm()
-                    move._action_assign()
-            if new_move:
-                self.run_procurement(
-                    new_move, qty_to_procure, mto_with_no_move_dest_id)
```
Comment (LoisRForgeFlow) [inline mrp_mto_with_stock/models/mrp_production.py]: why did you remove all this logic?

File: mrp_mto_with_stock/models/mrp_production.py [hunk 9]
```diff
@@ -122,28 +119,21 @@ def run_procurement(self, move, qty, mto_with_no_move_dest_id):
...
         return True
 
-    @api.model
-    def _get_incoming_qty_waiting_validation(self, move):
-        """
-        This method should be overriden in submodule to manage cases where
-        we need to add quantities to the forecast quantity. Like draft
-        purchase order, purchase request, etc...
-        """
-        return 0.0
```
Comment (LoisRForgeFlow) [inline mrp_mto_with_stock/models/mrp_production.py]: why do you remove this hook?

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Code review

File: mrp_mto_with_stock/models/mrp_production.py (hunk 9 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: There is no more a submodule that needs this logic.
Anyway, any future module which would need this, can add this function.

File: mrp_mto_with_stock/__manifest__.py (hunk 7 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/__manifest__.py]: done

File: mrp_mto_with_stock/models/mrp_production.py (hunk 8 above)
Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: In the current version, this logic create a duplication of the procurement

File: mrp_mto_with_stock/models/mrp_production.py (hunk 9 above)
Comment (LoisRForgeFlow) [reply mrp_mto_with_stock/models/mrp_production.py]: The module `mrp_mto_with_stock_purchase` was using it, and the idea of having this here is that you can extend that logic in different submoules and they would be able to coexist or be installed individually.

Comment (sergiocorato) [reply mrp_mto_with_stock/models/mrp_production.py]: The module `mrp_mto_with_stock_purchase` is no more needed and won't be migrated, if you think this is needed please explain why and migrate it.
Anyway this is dead code until someone add a module that uses it, and in that possible future PR that add this module, that code can be added too, without problem.

Comment (LoisRForgeFlow) [reply mrp_mto_with_stock/models/mrp_production.py]: Ok, fair enough

**Comment by LoisRForgeFlow** [comment]
@sergiocorato One more last thing, can you reword your last commit to something more meaninful? Thanks!

image

**Comment by sergiocorato** [comment]
> @sergiocorato One more last thing, can you reword your last commit to something more meaninful? Thanks!
> 
> image

@LoisRForgeFlow done, thanks

**Review by LoisRForgeFlow — APPROVED** [review]

**Comment by sergiocorato** [comment]
@rousseldenis @kos94ok-3D @andreampiovesana @LoisRForgeFlow @vipvalen @desdelinux @raistdev  @OCA/manufacturing-maintainers 

I found a bug that tests not detected and as I'm not able to test without purchase, I created #707 where I added depends on `purchase_mrp` and `stock_mts_mto_rule` to do some more complex and real tests.

If these depends are undesirable, I'll create another PR with only the changes in code (or someone help me to create test without purchase 8)).
